cfep-13.md

```diff
@@ -0,0 +1,64 @@
+
+<table>
+<tr><td> Title </td><td> Securing `conda-forge` Uploads to `anaconda.org` </td>
+<tr><td> Status </td><td> Proposed </td></tr>
+<tr><td> Author(s) </td><td> Matt, Marius, Anthony, Isuru</td></tr>
+<tr><td> Created </td><td> Feb 27, 2020</td></tr>
+<tr><td> Updated </td><td> Feb 27, 2020</td></tr>
+<tr><td> Discussion </td><td> NA </td></tr>
+<tr><td> Implementation </td><td> NA </td></tr>
+</table>
+
+## Abstract
+
+Currently the `anaconda.org` tokens on each feedstock allow any `conda-forge`
+feedstock maintainer to upload any output to any `anaconda.org` package.
+This could allow someone to push a compromised and/or malicious build
+of say the `python` package. This situation is a security risk that we
+should mitigate.
+
+## Specification
+
+We will secure the uploads of outputs to `anaconda.org` via the following process.
+
+1. Each feedstock is provisioned with a secret key that is only known to the
+   maintainers of that feedstock and the core team. These keys will be kept in
+   a private repo with only core members having write access. This key will be pushed to
+   the CI providers for that repo as well.
+2. Feedstock creation will move to a separate process in order to secure the keys
+   associated with our CI provider accounts and to securely register outputs of feedstocks.
+3. A public, global registry of all outputs from all feedstocks will be kept
+   in a repo on github. Write access to this repo is limited to core as well.
+3. When a feedstock builds new outputs it wants to upload, it will push these outputs
+   to a staging organization on `anaconda.org`. It will also send the names of the
+   outputs, their SHA256 checksums, the feedstock name, and the secret key to a
+   validation web service. This web service then verifies the secret key-feedstock
+   combination and verifies that the outputs are allowed for that feedstock.
+4. If everything is ok, the outputs are copied from the staging organization to the
+   main conda-forge channels.
+6. The service should post a comment on the PR that the outputs were copied.
```

Comment:
How about just waiting for the webservice to report that the package was copied?

Reply:
Not sure what you mean. Once the web service has copied, it should report back to let people know.

Reply:
In CI, we send a request to the webservice and the webservice can send a 200 code if the package was copied and a 403 if it's forbidden. CI could fail if the code is not 200.

Reply:
Ah I see. That is fine for the CI, but it doesn't really help people understand why something failed. I'd like things to be transparent enough so that users know which thing to go track down when we inevitably don't copy something due to a bug, etc.